### PR TITLE
Enabled stronk types as non-type template params

### PR DESCRIPTION
**Pull request by anders-wind**

### CODE REVIEW

File: include/stronk/stronk.h [hunk 1]
```diff
@@ -6,62 +6,78 @@
...
+
 template<typename Tag, typename T, template<typename> typename... Skills>
 struct stronk : public Skills<Tag>...
 {
-  private:
-    T _v;
-
-  public:
     using underlying_type = T;
+    // we need this value to be publicly available for stronk types to be usable as non-type template parameters.
```
Comment (engesgaard) [inline include/stronk/stronk.h:22]: Is it possible to explain why this is required to use non-type template parameters? Or does that become too complicated?

File: include/stronk/stronk.h [hunk 2]
```diff
@@ -6,62 +6,78 @@
...
 
 #include <stronk/utilities/equality.h>
+#include <stronk/utilities/macros.h>
 #include <stronk/utilities/strings.h>
 
 namespace twig
 {
+
+template<typename T>
+concept should_be_copy_constructed = std::is_trivially_copyable_v<T> && sizeof(T) <= sizeof(T*);
```
Comment (engesgaard) [inline include/stronk/stronk.h:16]: Would `sizeof(void*)` make it clearer that this is just the pointer size on the system? If I have understood the code correctly

File: include/stronk/stronk.h [hunk 3]
```diff
@@ -6,62 +6,78 @@
...
 struct stronk : public Skills<Tag>...
 {
-  private:
-    T _v;
-
-  public:
     using underlying_type = T;
+    // we need this value to be publicly available for stronk types to be usable as non-type template parameters.
+    // Therefore, to discourage direct usage of it, we have given it a long ugly name.
+    T _you_should_no_be_using_this_but_rather_unwrap;
```
Comment (engesgaard) [inline include/stronk/stronk.h:24]: no -> not

File: tests/src/stronk_tests.cpp [hunk 4]
```diff
@@ -533,4 +533,15 @@ TEST(can_index, can_index_works_for_vectors)
...
+struct a_none_type_template_parameter_type : stronk<a_none_type_template_parameter_type, int32_t>
+{
+    using stronk::stronk;
+};
+template<a_none_type_template_parameter_type Val>
+struct AnotherType
+{
+};
+
+constexpr static auto gravity = AnotherType<a_none_type_template_parameter_type {25}>();  // NOLINT
```
Comment (engesgaard) [inline tests/src/stronk_tests.cpp:545]: Very high gravity in stronk land :-D

File: benchmarks/src/benchmark_helpers.h [hunk 5]
```diff
@@ -25,42 +29,84 @@ struct string_wrapping_type : twig::stronk_default_unit<string_wrapping_type, st
...
+    return rng;
+}
+
+template<typename T>
+auto rand(T mi, T ma) -> T
+{
+    if constexpr (std::is_floating_point_v<T>) {
+        auto dist = std::uniform_real_distribution<T>(mi, ma);
+        return dist(get_random_device());
+    } else if constexpr (std::is_integral_v<T> && sizeof(T) <= 1) {
```
[the remark this answers is not in the document]
Comment (anders-wind) [reply benchmarks/src/benchmark_helpers.h:57]: char, and uint8_t does not have a int distribution class implementation on msvc

File: benchmarks/src/construction_benchmarks.cpp [hunk 6]
```diff
@@ -6,51 +6,82 @@
...
+
+template<typename T>
+static void benchmark_copy_vector_of(benchmark::State& state)
+{
+    auto vec = std::vector<T>(static_cast<size_t>(state.range(0)));
+    std::generate(vec.begin(), vec.end(), []() { return generate_randomish<T> {}(); });
+    benchmark::DoNotOptimize(vec.data());
+    for (auto _ : state) {
+        auto copy_into = std::vector<T>();
+        benchmark::DoNotOptimize(copy_into.data());
```
[the remark this answers is not in the document]
Comment (anders-wind) [reply benchmarks/src/construction_benchmarks.cpp:60]: Ill simplify

[the remark this answers is not in the document]
Comment (anders-wind) [reply benchmarks/src/construction_benchmarks.cpp:60]: Should not affect anything though

**Review by engesgaard — APPROVED** [review]

**Pull request merged**